piptools/utils.py

```diff
@@ -88,6 +93,22 @@ def comment(text: str) -> str:
     return click.style(text, fg="green")
 
 
+def canonicalize_ireq(ireq: InstallRequirement) -> InstallRequirement:
+    """
+    Return a copy of ireq with canonicalized extras strings
+    """
+    ireq = copy_install_requirement(
+        ireq, extras=set(map(canonicalize_name, ireq.extras))
```

Comment:
> > `copy_install_requirement`
>
> Do you think this logic belongs in or out of `copy_install_requirement`?

That sounds reasonable. Let's move this normalization to copy_install_requirement and use this function in all our InstallRequirement constructors.